xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/doc/XWikiDocument.java

```diff
@@ -3446,11 +3457,150 @@ public void readObjectsFromForm(EditForm eform, XWikiContext context) throws XWi
+                    continue;
+                }
+                classNumber = Integer.parseInt(classNumberAsString);
+            } catch (XWikiException e) {
+                // If the class page cannot be found, skip the property update
+                LOGGER.warn("Failed to load document [{}], ignoring property update [{}]. Reason: [{}]",
+                    classReference, parameter.getKey(), ExceptionUtils.getRootCauseMessage(e));
+                continue;
+            } catch (NumberFormatException e) {
+                // If the numner isn't valid, skip the property update
+                LOGGER.warn("Invalid xobject number [{}], ignoring property update [{}].", classNumberAsString,
+                    parameter.getKey());
+                continue;
+            }
+            SortedMap<Integer, Map<String, String[]>> objectMap = result.get(className);
+            if (objectMap == null) {
+                objectMap = new TreeMap<>();
+                result.put(className, objectMap);
+            }
+            // Get the property from the right object #objectNumber of type 'objectName'; create it if they don't exist
+            Map<String, String[]> object = objectMap.get(classNumber);
+            if (objectMap.get(classNumber) == null) {
+                object = new HashMap<>();
+                objectMap.put(classNumber, object);
+            }
+            object.put(classPropertyName, parameter.getValue());
+        }
+        return result;
+    }
+
+    /**
+     * Create and/or update objects in a document given a list of HTTP parameters of the form
+     * '<spacename>.<classname>_<number>_<propertyname>'. If the object already exists, the field is replace by the
+     * given value. If the object doesn't exist in the document, it is created then the property <propertyname> is
+     * initialized with the given value. An object is only created if the given '<number>' is 'one-more' than the
+     * existing number of objects. For example, if the document already has 2 objects of type 'Space.Class', then it
+     * will create a new object only with 'Space.Class_2_prop=something'. Every other parameter like
+     * 'Space.Class_42_prop=foobar' for example, will be ignore.
+     * 
+     * @param eform Form information that contains all the query parameters
+     * @param context
+     * @throws XWikiException
+     * @since 7.0RC1
+     */
+    public void readObjectsFromFormUpdateOrCreate(EditForm eform, XWikiContext context) throws XWikiException
+    {
+        Map<String, SortedMap<Integer, Map<String, String[]>>> fromRequest =
+            parseRequestUpdateOrCreate(eform.getRequest(), context);
+        for (Entry<String, SortedMap<Integer, Map<String, String[]>>> requestClassEntries : fromRequest.entrySet()) {
+            DocumentReference requestClassReference =
+                getCurrentDocumentReferenceResolver().resolve(requestClassEntries.getKey());
```

Comment:
As mentioned above, if your key is a DocumentRefence, you don`t need to resolve again the same class name string.